Add ReplaceLineEndings and EnumerateLines

GrabYourPitchforks opened the pull request:
Resolves https://github.com/dotnet/runtime/issues/40315.

Note to reviewers - the API surface, devdoc, and general behavior are complete. However, the implementation is naïve, as it's just a simple iterator instead of SIMD-optimized. The reason for this is that there's currently no overload of _IndexOfAny_ that fulfills the algorithmic complexity requirements we have.

Once https://github.com/dotnet/runtime/issues/49397 comes online it should be fairly straightforward to SIMD-optimize this without introducing layers of ISA special-casing or littering the workhorse method with unsafe code. But for now, let's get the basic functionality out and worry about optimizing it later.

GrabYourPitchforks committed:
Add ReplaceLineEndings and EnumerateLines

GrabYourPitchforks commented:
I need to send a follow-up iteration which removes some of the duplicated code and calls these new APIs instead.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs:
```diff
@@ -1152,6 +1152,95 @@ private string ReplaceHelper(int oldValueLength, string newValue, ReadOnlySpan<i
+            ReadOnlySpan<char> firstSegment = this.AsSpan(0, idxOfFirstNewlineChar);
+
+            ValueStringBuilder builder = new ValueStringBuilder(stackalloc char[256]);
+            foreach (ReadOnlySpan<char> nextSegment in this.AsSpan(idxOfFirstNewlineChar + charsToConsume).EnumerateLines())
```
>this.AsSpan(idxOfFirstNewlineChar + charsToConsume) [](http://example.com/codeflow?start=55&length=51)

interesting, I didn't know that `string.AsSpan(string.Length)` doesn't fail and it looks it return empty span. I am seeing you already having a test covering such case e..g "<CR>" case.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Text/NewLineUtility.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace System.Text
+{
+    internal static class NewLineUtility
```
>NewLineUtility [](http://example.com/codeflow?start=26&length=14)

why we call it New Line? why not just Line? just curious no real objection :-)

GrabYourPitchforks replied:
Deleted this file and moved it into string.cs as an internal helper. Also updated the comments to more clearly reflect that the Unicode Standard refers to these as "newline chars" and "newline functions". Hopefully this reads a bit more cleanly now!

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Text/NewLineUtility.cs:
```diff
@@ -0,0 +1,51 @@
+
+            for (int i = 0; i < span.Length; i++)
+            {
+                switch (span[i])
```
> switch (span[i]) [](http://example.com/codeflow?start=15&length=17)

Why you didn't include the vertical tab character too? 

U+000B <control-000B> (LINE TABULATION, vertical tabulation (VT))?

GrabYourPitchforks replied:
Per other fork, exclusion is intentional. I'll add a comment clarifying this.

tarekgh commented on src/libraries/System.Runtime/tests/System/StringTests.cs:
```diff
@@ -508,6 +507,68 @@ public static void EnumerateRunes(char[] chars, int[] expected)
+
+                return input.Replace("<CR>", "\r")
+                    .Replace("<LF>", "\n")
+                    .Replace("<VT>", "\v")
```
> .Replace("<VT>", "\v") [](http://example.com/codeflow?start=19&length=23)

interesting you have the VT test while you didn't include it in the parsing code.

GrabYourPitchforks replied:
It's a negative test to make sure that we _don't_ break on VT. The text of Unicode Sec. 5.8 goes into a little bit of history behind why this is. But the gist of it is that no platform ever standardized on VT as a newline function (NLF).

> Once NLF was reinterpreted to stand for a paragraph separator, in some cases another control character was pressed into service as a line separator. For example, vertical tabulation VT is used in Microsoft Word. However, the choice of character for line separator is even less standardized than the choice of character for NLF.

Per Recommendation R4 in Sec. 5.8, they list the sequences to be searched for as { NLF, LS, PS, FF }. Table 5-2 defines NLF as the set { CR, LF, CRLF, NEL }. So that brings the total set of sequences we're searching for as { CR, LF, CRLF, NEL, LS, PS, FF }. VT is specifically excluded from this list.

tarekgh replied:
ok, thanks for the explanation. do you think is worth it to add some comment why VT is not included? I know you have a comment talking about R4 Sec. 5.8 but the comment not mentioning VT.

stephentoub commented on src/libraries/System.Private.CoreLib/src/System/Text/NewLineUtility.cs:
```diff
@@ -0,0 +1,51 @@
+            // Importantly, this means that we cannot call MemoryExtensions.IndexOfAny, as that method
+            // is worst-case bounded by O(span.Length), not O(i). This means that when called in a loop,
+            // the worst-case runtime for the loop is O(span.Length ^ 2), which could lead to an algorithmic
+            // complexity attack.
```
I don't understand this.  Can you elaborate?  We will always jump forward to whatever is the result of IndexOfAny, so we'd never end up looking at any position more than once.  How would this degrade to O(n^2)?

GrabYourPitchforks replied:
The call to `IndexOfAny` may be _O(n)_, regardless of where the index actually occurs within the input. Imagine an input crafted such that there's a match at every index. So our first call to `IndexOfAny` will be _O(span.Length)_. Our second call might be _O(span.Length - 1)_. The third call might be _O(span.Length - 2)_. And so on.

The amortized cost between slices is _O(n)_, which puts the aggregate operation at _O(n^2)_.

stephentoub replied:
> The call to IndexOfAny may be O(n), regardless of where the index actually occurs within the input.

Why?  You're saying we might find the match at position 1 but still examine 1,000,000 more characters in the input?

GrabYourPitchforks replied:
> You're saying we might find the match at position 1 but still examine 1,000,000 more characters in the input?

`IndexOfAny` (for inputs > 5 elements) has this behavior today, yes. In fact, if you replace this routine with a call to `IndexOfAny`, the unit test _EnumerateLines\_EnumerationIsNotPolynomialComplexity_ (introduced as part of this PR) will detect the _O(n^2)_ operation and fail.

We have a few options for addressing this. We can change `IndexOfAny` to no longer have _O(n)_ behavior (which may severely harm perf for other common use cases). We can yet another special-cased code path within `IndexOfAny`. These specialized code paths within `IndexOfAny` tend to be very complex and utilize significant unsafe code. I had planned on addressing the performance issue within `NewLineUtility` directly using a minimum amount of unsafe code and relying on general vectorized APIs.

stephentoub replied:
> IndexOfAny (for inputs > 5 elements) has this behavior today, yes

Sorry, I feel like I'm missing something, as what you describe sounds unshippable (yet we apparently shipped it).  You're saying that even after IndexOfAny finds a match, it continues searching the entire rest of the input for another match?

If yes, we need to fix that.

stephentoub replied:
I just looked at the implementation:
https://github.com/dotnet/runtime/blob/ebba45155160c7056faf1675448af1b75e9f4a48/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs#L597-L609
Seems the behavior you describe is in fact there and happens because we're doing a naive "first search for value 1, then if it couldn't be found search for value 2, then if it couldn't be found search for value 3, ... and so on".  Ugh.

GrabYourPitchforks replied:
> IndexOfAny is used by a non-trivial amount of code to do parsing like this PR is doing. The workaround employed of not using IndexOfAny doesn't help those other uses. I think we need to fix this.

I agree that we should consider a change here. The reason I wasn't pushing the issue hard is that I couldn't find any perf-critical call sites within our SDK which provide 6+ needles. The highest we've gone in the past is 5 needles (for `File` APIs), and we have a specialized code path to handle that. As far as I know, this PR introduces the first API which needs to perform a 6-needle search.

My main concern behind making a change to `IndexOfAny` for the 6+ needle case is that it could be a significant perf regression in the case where the needles are sorted roughly in order of likelihood. We can guarantee _O(i * p)_ runtime \[where _i_ is the index of the first match in the haystack and _p_ is the number of needles being sought\], but I wonder if that "_\* p_" part is going to irritate people.

GrabYourPitchforks replied:
Also, there is the whole overhead of setting up the SIMD loop on *ever single entrance* to `IndexOfAny`, but eh. Unless we often expect matches to occur near the beginning of the span this overhead might not be worth worrying about.

stephentoub replied:
Yeah, totally get there will be tradeoffs, and we may not like all of them.  But for a long search space, I think where we currently are is worse.  Thanks for thinking deeply about it and raising the issue.

GrabYourPitchforks committed:
Merge commit '5a1b8223dab2f7954e7f206095ba937e5d237299' into str_lineend

GrabYourPitchforks committed:
PR feedback, rely on MemoryExtensions.IndexOfAny

GrabYourPitchforks committed:
Update unit tests

GrabYourPitchforks commented:
### Changes in latest iteration

- Removed custom iterator logic from `NewLineUtility` now that https://github.com/dotnet/runtime/pull/53652 is checked in, and just riding directly atop `IndexOfAny`
- Small optimization in `string.ReplaceNewLines` to avoid two memcpy operations
- Updated unit tests which had their own custom line split logic and had them ride atop the newly introduced methods

This last item in particular simplifies some of the unit test logic (particularly noticeable for _System.Data_). I only updated unit tests that appeared that they __were not__ testing newline behavior as part of a protocol. For example, RFC 2616 (HTTP/1.1) gives a specific definition of newline chars which is incompatible with the Unicode's definition of newline chars, so I didn't update the _System.Net.Http_ unit tests to use the new helper.

__Area owners, please verify that the changes I made to your unit tests are valid.__
(/cc a bunch of people)

* _System.Data_: @ajcvickers @cheenamalhotra @david-engel
* _System.Text.Json_: @eiriktsarpalis @layomia @steveharter
* _System.Xml_: @buyaa-n @joperezr @krwq
* _System.Net.NetworkInformation_: @dotnet/ncl

krwq commented on src/libraries/System.Private.Xml.Linq/tests/TreeManipulation/SaveWithFileName.cs:
```diff
@@ -256,11 +256,6 @@ private static string StripOffXmlDeclaration(string s)
-            s = s.Replace("\r", "");
-            return s;
-        }
+        private static string NormalizeNewLines(string s) => s.ReplaceLineEndings("");
```
Would `s.ReplaceLineEndings()` work the same?

krwq replied:
nvm, I just read the docs says it will replace with Environment.NewLine but I personally don't find it particularly obvious (I was expecting to remove them)

krwq reviewed (approved):
XML part looks good to me

ManickaP reviewed (approved):
NetworkInformation LGTM

eiriktsarpalis reviewed (approved):
:shipit: from the json perspective

GrabYourPitchforks commented:
Thanks all for the input! I'm going ahead and committing this for now, assuming everything is ok.

__System.Data__ crew (@ajcvickers @cheenamalhotra @david-engel), if this feels inappropriate then please ping me, or open a new tracking issue, or simply revert the relevant portions of this commit. Whatever is easiest for you all. :)

David-Engel commented:
> if this feels inappropriate then please ping me, or open a new tracking issue, or simply revert the relevant portions of this commit. Whatever is easiest for you all. :)

@GrabYourPitchForks Looks good to me. Plus the new code is so much cleaner! :)
